### ISSUE TITLE
Processing GPUImageVideoCamera as CMSampleBufferRef for QuickBlox SDK

### ISSUE THREAD

**Issue by rjburdish** [BradLarson/GPUImage#1763]
I am currently using the VideoChat functionality of Quickblox SDK. Following the Custom Capture Session tutorial, I am able to create a custom video chat session between two users. However, I would like to apply filters to the stream before sending/forwarding the stream to quickblox.

Using an AVCaptureSession I create a serial queue to handle the processing of the frames with the following code:

dispatch_queue_t callbackQueue= dispatch_queue_create("cameraQueue", NULL);
[videoCaptureOutput setSampleBufferDelegate:self queue:callbackQueue];
Which then calls the function:
- (void)captureOutput:(AVCaptureOutput *)captureOutput  didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
  [self.videoChat processVideoChatCaptureVideoSample:sampleBuffer];
  }
  Where 'self.videoChat' is of type QBVideoChat. I would like to take the output of 'GPUImageVideoCamera' and pass it through a filter, and finally forward the processed video to quickblox.

So far I can get the filter working by adding a target to the 'GPUImageVideoCamera', then adding a 'GPUImageView' as the target to the processed video as so:

[_liveVideo addTarget:myFilter];
[myFilter addTarget:_ImageView];
[_liveVideo startCameraCapture];
So, in conclusion, I cannot seem to figure out how to get the output from 'myFilter' forwarded to QuickBlox videoChat and was hoping someone can point me in the right direction. I hope this question was descriptive enough, and thanks in advance.

**Comment by eXhausted** [comment]
Do we need lockFramebufferForReading and unlockFramebufferAfterReading?

Cheers.

**Comment by rued** [comment]
Not sure, when I did it, it was before the frame caching system.

**Comment by omarojo** [comment]
@rued Hi, thanks for your code example, although I notice you set the SampleTimingInfo using your [SMFilterManager sharedInstance]-> sampleTimimgInfo. Do you know how this parameter works ? or how do I set it up ? is it a constant or what?

**Comment by rued** [comment]
I used CMSampleBufferGetSampleTimingInfo on the original CMSampleBuffer... however later on I removed that and just used a fresh zero time:

**Comment by omarojo** [comment]
@rued well Im actually looking looking for a service to broadcast video (No P2P, nor video chat). Im now experimenting with Kickflip.io but their support is less to none. And digging in their library, I see they append a SampleBuffer to an AssetWriterInput when capturing from the Camera. 

Im trying to substitute the native camera output buffer (```- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection;``` )
with GPUImage sample buffer using the block method mentioned in this thread.

But I keep getting error when the AssetWriter tries to append the customized sample buffer I created.
```Error Domain=AVFoundationErrorDomain Code=-11800 "The operation could not be completed" UserInfo={NSUnderlyingError=0x14b3328a0 {Error Domain=NSOSStatusErrorDomain Code=-12780 "(null)"}, NSLocalizedFailureReason=An unknown error occurred (-12780), NSLocalizedDescription=The operation could not be completed}```

by the way, I set the timingInfo to ```            CMSampleTimingInfo timimgInfo =  {kCMTimeInvalid, kCMTimeZero, kCMTimeInvalid};```

no luck :(
If you know a good video broadcasting service with mobileSDK, that would be appreciated.
